exercises/Toluwalase_Agoro/ex02_oo_basics/src/ShapeSorter.cpp

```diff
@@ -0,0 +1,123 @@
+#include <iostream>
+#include <memory>
+#include <iomanip>
+#include <string>
+#include "BaseShape.h"
+#include "Triangle.h"
+#include "Rectangle.h"
+#include "Square.h"
+#include "Circle.h"
+#include "ShapeSorter.h"
+
+
+
+void ShapeSorter::addShape(std::shared_ptr<BaseShape> newShape) {
+  newShape-> calculateArea();
+  newShape-> calculatePerimeter();
+  m_allShapes.emplace_back(newShape);
+
+}
+
+void ShapeSorter::displayByType(std::string& shapeType) {
+  int i{ 1 };
+  for (auto& shape : m_allShapes) {
+    if (shape->m_type == shapeType) {
+
+      shape->calculateArea();
+      shape->calculatePerimeter();
+
+      std::cout << i << ".";
+      std::cout << std::left << std::setw(10) << shape->m_type;
+      std::cout << std::setprecision(0);
+      std::cout << std::setw(20) << "Number of sides: " << shape->m_numberOfSides << "  ";
+      std::cout << std::fixed << std::setprecision(3);
+      std::cout << std::setw(8) << "Area: " << shape->m_area << "  ";
+      std::cout << std::setw(15) << "Perimeter: " << shape->m_perimeter << std::endl;
+      i++;
+    }
+  }
+}
+
+void ShapeSorter::displayByArea() {
+
+  if (m_orderToReadArea.size() == 0) {
+    for (int i = 0; i < m_allShapes.size(); i++) {
+      double maxArea{ 0 };
+      int index{ 0 };
+      for (int j = 0; j < m_allShapes.size(); j++) {
+        if (m_allShapes[j]->m_area > maxArea && std::find(m_orderToReadArea.begin(), m_orderToReadArea.end(), j) == m_orderToReadArea.end()) {
+          index = j;
+          maxArea = m_allShapes[j]->m_area;
+        }
```

Comment:
Have you come across the [sort](http://www.cplusplus.com/reference/algorithm/sort/) function in the standard library? This can be used here to do most of the work for us.

Reply:
Same for perimeter and number of sides